runtime/tr.source/trj9/z/codegen/ReduceSynchronizedFieldLoad.cpp

```diff
@@ -66,71 +70,112 @@ ReduceSynchronizedFieldLoad::inlineSynchronizedFieldLoad(TR::Node* node, TR::Cod
 
-   // Sanity check for instruction kind as we are doing an unsafe cast following this
-   TR_ASSERT_SAFE_FATAL(loadInstruction->getKind() == OMR::Instruction::Kind::IsRX || loadInstruction->getKind() == OMR::Instruction::Kind::IsRXY, "Expecting the append instruction to be a load of kind RX or RXY\n");
+   TR_ASSERT_SAFE_FATAL(loadInstruction->isLoad() && (loadInstruction->getKind() == OMR::Instruction::Kind::IsRX || loadInstruction->getKind() == OMR::Instruction::Kind::IsRXY), "Expecting the append instruction to be a load of kind RX or RXY\n");
 
    TR::MemoryReference* loadMemoryReference = static_cast<TR::S390RXInstruction*>(loadInstruction)->getMemoryReference();
 
    TR_ASSERT_SAFE_FATAL(loadMemoryReference->getIndexRegister() == NULL, "Load memory reference must not have an index register\n");
 
-   int32_t lockWordOffset = node->getChild(2)->getConst<int32_t>();
+   // Recreate the memory reference since the load instruction will be moved to the OOL code section and the register
+   // allocated to hold value of the base register of the memory reference may not be valid in the main line path
+   loadMemoryReference = generateS390MemoryReference(*loadMemoryReference, 0, cg);
+
+   TR::Node* lockWordOffsetNode = node->getChild(2);
+
+   int32_t lockWordOffset = lockWordOffsetNode->getConst<int32_t>();
 
    TR::MemoryReference* lockMemoryReference = generateS390MemoryReference(objectRegister, lockWordOffset, cg);
 
    const bool generateCompressedLockWord = static_cast<TR_J9VMBase*>(cg->comp()->fe())->generateCompressedLockWord();
 
-   const bool generate32BitLoads = TR::Compiler->target.is32Bit() || (TR::Compiler->target.is64Bit() && generateCompressedLockWord && node->getChild(1)->getOpCodeValue() == TR::iloadi);
+   const bool is32BitLock = TR::Compiler->target.is32Bit() || generateCompressedLockWord;
+   const bool is32BitLoad = loadNode->getOpCodeValue() == TR::l2a || loadNode->getSymbolReference()->getSymbol()->getSize() == 4;
 
-   if (generate32BitLoads)
+   if (is32BitLock && is32BitLoad)
       {
-      TR::MemoryReference* reusedLoadMemoryReference = generateS390MemoryReference(*loadMemoryReference, 0, cg);
-
-      generateSSFInstruction(cg, TR::InstOpCode::LPD, node, registerPair, reusedLoadMemoryReference, lockMemoryReference);
+      generateSSFInstruction(cg, TR::InstOpCode::LPD, node, registerPair, loadMemoryReference, lockMemoryReference);
 
-      if (node->getChild(1)->getOpCodeValue() == TR::l2a)
+      if (loadNode->getOpCodeValue() == TR::l2a)
          {
          generateRREInstruction(cg, TR::InstOpCode::LLGFR, node, loadRegister, loadRegister);
          }
       }
    else
       {
-      TR_ASSERT_SAFE_FATAL((lockWordOffset & 7) == 0 || (generateCompressedLockWord && (lockWordOffset & 3) == 0),
-            "Lockword must be aligned on a double-word boundary or on a word boundary if the lockword is compressed\n");
+      // LPDG requires memory references to be aligned to a double-wrod boundary
```

Comment:
spelling nit.  `double-wrod`->`double-word`

Reply:
Fixed in 98f8017